doc/adr006-modules.lit.md

```diff
@@ -0,0 +1,553 @@
+      fdec["4"]
+      finc["8"]
+      i1["1"]
+      i2["5"]
+      i3["10"]
+      i4["11"]
+    end
+
+    subgraph "Intermediate representation"
+      dec["def 'dec': { id: 4, ... }"]
+      inc["def 'inc': { id: 8, ... }"]
+      dec_i["param 'i': { id: 10, ...}"]
+      inc_i["param 'i': { id: 11, ...}"]
+    end
+
+    finc --> inc
+    fdec --> dec
+    i1 --> dec_i
+    i2 --> inc_i
+    i3 --> dec_i
+    i4 --> inc_i
+```
+
+Now, let's have a look at the module `general`:
+
+```bluespec generated/adr006after.qnt +=
+module general {
+  import fun.*
+
+  const N: int
+  var x: int
+
+  action init = { x' = N }
+  action step = { x' = dec(x) }
+}
+```
+
+As you see, we do not change anything in the syntax. Similar to the table
+of `fun`, we organize the lookup table of `general` as follows:
+
+```mermaid
+graph TB
+    subgraph "table for 'general'"
+      fdec["4"]
+      finc["8"]
+      i1["1"]
+      i2["5"]
+      i3["10"]
+      i4["11"]
+      gN["13"]
+      gx["15"]
+      ...
+    end
+
+    subgraph "Intermediate representation"
+      dec["def 'dec': { id: 4, ... }"]
+      inc["def 'inc': { id: 8, ... }"]
+      dec_i["param 'i': { id: 10, ...}"]
+      inc_i["param 'i': { id: 11, ...}"]
+      N["const 'N': { id: 13 }"]
+      x["const 'x': { id: 15 }"]
+      ...
+    end
+
+    finc --> inc
+    fdec --> dec
+    i1 --> dec_i
+    i2 --> inc_i
+    i3 --> dec_i
+    i4 --> inc_i
+    gN --> N
+    gx --> x
+```
+
+The interesting outcome of this approach is that we do not need a separate
+lookup table per module. Since we are using unique identifiers as keys, we
+could simply produce a flat table for all modules.
+
+### 4.2. Addressing the issue D3
+
+To address the need for treating variables of different instances as different
+objects, we propose to introduce a new flattening stage in the pipeline:
+
+```mermaid
+graph LR
+   repl["REPL"]
+   cli["CLI"] --> repl
+   cli --> loader["Source loader"] --> parser["Parser"]
+   repl --> loader
+   parser --> resolver["Name resolver"] --> analyzer
+   flattener["Flattener"]
+   analyzer --> flattener --> js
+   flattener --> apalache["Apalache transpiler"]
+
+   subgraph analyzer["Analyzer"]
+     types["Type inferrer"]
+     effects["Effect inferrer"]
+     updates["Updates checker"]
+     modes["Modes checker"]
+     types --> effects --> updates --> modes
+   end
+
+   subgraph runtime["JS runtime"]
+     js["JS transpiler"]
+     simulator["Simulator"]
+     testing["Unit testing"]
+
+     js --> simulator --> testing
+   end
+```
+
+The job of the flattener would be to clone all definitions of the instantiated
+module and tune them with the supplied expressions for the constants.
+Effectively, the flattener would transform the module `fixed` of our example
+into the following code:
+
+```bluespec generated/adr006instances.qnt +=
+module fixed {
+  // the instance general(N = 3)
+
+  // this is how `import fun.*` is transformed
+  import fun.* as I3
```

Comment:
These lines are unclear to me. Are we replacing `include fun.* as I3` with `import fun.* as I3`? Seems like we should be omitting this last line in flattening.

Reply:
This is just how the translation of `import fun.*` could look like, if we did flattening by code transformation. In principle, this could all be done via the lookup table.